api/include/opentelemetry/trace/tracer.h

```diff
@@ -163,6 +163,18 @@ class Tracer
     }
   }
 
+#if OPENTELEMETRY_ABI_VERSION_NO >= 2
+  /**
+   * Reports if the tracer is enabled or not. A disabled tracer will not create spans.
+   *
+   * The instrumentation authors must call this method very time before creating a span to
```

Comment:
```suggestion
   * The instrumentation authors must call this method every time before creating a span to
```

Reply:
I don't think this constraint is mandatory according to the specifications. Ideally, instrumentation authors should have the flexibility to decide when to call the Enabled() method - particularly in cases where creating a span would require expensive pre-computation operations.

Reply:
Good catch, this is not mandatory, only a helper for expensive operations.

@psx95 Please reword.

Reply:
Also if the purpose is avoid heavy computation for the user to call this `Enabled()`, I think the call to `Enabled()` involves virtual dispatch which seems not a light operation. 

Could we save the enabled in the API class so enabled check just query this class member? And also provide API for SDK to set it?

Reply:
+1 I think @ThomsonTan implemented something similar for Log SDK, good to refer that too.

Reply:
I think I misunderstood [the spec here](https://github.com/open-telemetry/opentelemetry-specification/blame/8bbac83c0c994c0a750cc688935fc1cae7fd39b3/specification/trace/api.md#L216), I will reword it, I agree instrumentation author should have flexibility on when to call this 👍🏻

Reply:
Regarding avoiding the virtual dispatch call: I'm not sure if the associated overhead is huge to move implementation to the API (I might be wrong here, so please let me know). 

Unlike the Logger, in which the implementation was completely moved to the API in a self-sufficient way, it might be difficult to provide an accurate implementation in API itself. 
As @ThomsonTan mentioned, we would need to add a protected method in the API which the SDKs can call to update the value of the saved variable. 

So we won't have a way to force implementations of the API to call the `UpdateEnabled(bool enabled)` method. If the SDK implementations forget to call this method, the API might return incorrect state of the tracers.
Its a minor concern, bur are we ok with this approach ?

Reply:
@ThomsonTan I have implemented your suggestion in [this commit](https://github.com/open-telemetry/opentelemetry-cpp/pull/3357/commits/8037eaff29376f9442a0a62322501535d99ea771). 

 - Pros: It avoids the virtual methods that lead to dynamic dispatch calls.
 - Cons: It forces the SDK implementations to explicitly call `UpdateEnabled(bool)` to keep status up-to-date.

On second look, it seems like a good trade-off to me. Let me know your thoughts on this.